cli/command/manifest/fetch.go

```diff
@@ -0,0 +1,342 @@
+			}
+			continue
+		}
+
+		if confirmedV2 && endpoint.Version == registry.APIVersion1 {
+			logrus.Debugf("Skipping v1 endpoint %s because v2 registry was detected", endpoint.URL)
+			continue
+		}
+
+		if endpoint.URL.Scheme != "https" {
+			if _, confirmedTLS := confirmedTLSRegistries[endpoint.URL.Host]; confirmedTLS {
+				logrus.Debugf("Skipping non-TLS endpoint %s for host/port that appears to use TLS", endpoint.URL)
+				continue
+			}
+		}
+
+		logrus.Debugf("Trying to fetch image manifest of %s repository from %s %s", normalName, endpoint.URL, endpoint.Version)
+
+		fetcher, err := newManifestFetcher(endpoint, repoInfo, authConfig, registryService)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+
+		if foundImages, err = fetcher.Fetch(ctx, namedRef); err != nil {
+			// Was this fetch cancelled? If so, don't try to fall back.
+			fallback := false
+			select {
+			case <-ctx.Done():
+			default:
+				if fallbackErr, ok := err.(fallbackError); ok {
+					fallback = true
+					confirmedV2 = confirmedV2 || fallbackErr.confirmedV2
+					if fallbackErr.transportOK && endpoint.URL.Scheme == "https" {
+						confirmedTLSRegistries[endpoint.URL.Host] = struct{}{}
+					}
+					err = fallbackErr.err
+				}
+			}
+			if fallback {
+				if _, ok := err.(distribution.ErrNoSupport); !ok {
+					// Because we found an error that's not ErrNoSupport, discard all subsequent ErrNoSupport errors.
+					discardNoSupportErrors = true
+					// save the current error
+					lastErr = err
+				} else if !discardNoSupportErrors {
+					// Save the ErrNoSupport error, because it's either the first error or all encountered errors
+					// were also ErrNoSupport errors.
+					lastErr = err
+				}
+				continue
+			}
+			logrus.Errorf("Not continuing with pull after error: %v", err)
+			return nil, nil, err
+		}
+
+		if transactionID == "" && len(foundImages) > 1 {
+			transactionID = normalName
+		}
+		// Additionally, we're never storing on inspect, so if we're asked to save images it's for a create,
+		// and this function will have been called for each image in the create. In that case we'll have an
+		// image name *and* a transaction ID. IOW, foundImages will be only one image.
+		if !fetchOnly {
+			if err := storeManifest(foundImages[0], makeFilesafeName(normalName), transactionID); err != nil {
+				logrus.Errorf("Error storing manifests: %s\n", err)
+			}
+		}
+		return foundImages, repoInfo, nil
+	}
+
+	if lastErr == nil {
+		lastErr = fmt.Errorf("no endpoints found for %s\n", normalName)
+	}
+
+	return nil, nil, lastErr
+
+}
+
+func newManifestFetcher(endpoint registry.APIEndpoint, repoInfo *registry.RepositoryInfo, authConfig types.AuthConfig, registryService registry.Service) (manifestFetcher, error) {
+	switch endpoint.Version {
+	case registry.APIVersion2:
+		return &v2ManifestFetcher{
+			endpoint:   endpoint,
+			authConfig: authConfig,
+			service:    registryService,
+			repoInfo:   repoInfo,
+		}, nil
+	case registry.APIVersion1:
```

Comment:
Maybe not clear in the initial comment, but, this subcommand isn't *just* for manifest lists. It also allows you to do `docker manifest inspect [blah]` of an image name that could be living in a v1 registry. This is so that users can do the "shallow pull" of an image that several people had requested in earlier issues